pkg/monitor/podipcontroller/pod_ip_controller.go

```diff
@@ -0,0 +1,203 @@
+		AddFunc: c.Enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.Enqueue(newObj)
+		},
+		DeleteFunc: c.Enqueue,
+	})
+
+	return c
+}
+
+func (c *SimultaneousPodIPController) Enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid queue key '%v': %v", obj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *SimultaneousPodIPController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	fmt.Printf("Starting SimultaneousPodIPController")
+	defer func() {
+		fmt.Printf("Shutting down SimultaneousPodIPController")
+		c.queue.ShutDown()
+		fmt.Printf("SimultaneousPodIPController shut down")
+	}()
+
+	if !cache.WaitForNamedCacheSync("SimultaneousPodIPController", ctx.Done(), c.cachesToSync...) {
+		return
+	}
+
+	go func() {
+		// this only works because we have a single thread consuming
+		wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}()
+
+	<-ctx.Done()
+}
+
+func (c *SimultaneousPodIPController) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *SimultaneousPodIPController) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(ctx, key.(string))
+
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %w", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *SimultaneousPodIPController) sync(ctx context.Context, key string) error {
+	//klog.V(4).Infof("Starting SimultaneousPodIPController sync")
+	//defer klog.V(4).Infof("SimultaneousPodIPController sync done")
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// only consider pod network pods because host network pods will have duplicated IPs
+	if pod.Spec.HostNetwork {
+		return nil
+	}
+	if isPodIPReleased(pod) {
+		return nil
+	}
+
+	otherPods, err := c.podLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	podLocator := monitorapi.LocatePod(pod)
+	podIPToPodNames := map[string]sets.String{}
+	for _, currIP := range pod.Status.PodIPs {
+		currPodIP := currIP.IP
+		podIPToPodNames[currPodIP] = sets.NewString(podLocator)
+
+		// iterate through every ip of every pod.  I wonder how badly this will scale
+		for _, otherPod := range otherPods {
+			if otherPod.Spec.HostNetwork {
+				continue
+			}
+			if otherPod.UID == pod.UID {
+				continue
+			}
+			if isPodIPReleased(otherPod) {
+				continue
+			}
+
+			for _, otherIP := range otherPod.Status.PodIPs {
+				otherPodIP := otherIP.IP
+				if currPodIP == otherPodIP {
```

Comment:
since you are going to iterate through all the pods anyway, you might as well ignore the key of the pod you get the event for and just check for duplicates across all pods. As you iterate, just store each IP as a map key with the value as the pod namespaced name. Then during iteration if you detect a collision record that one and so on.